### ISSUE TITLE
Memory leak when decompression fails with `ERROR_GENERIC`

### ISSUE THREAD

**Issue by jmg-duarte** [python-lz4/python-lz4#247]
We're using `lz4` in production and noticed some of our machines were crashing with OOM errors.

We found out that they were stuck in a loop where they tried to decompress a faulty file, the error provided by `lz4` was simply `ERROR_GENERIC`; the memory usage kept getting higher until the process/machine crashed with OOM.

Expected:
- The memory usage to *temporarily* increased (i.e. when decompressing) and get back to the initial after failing.

To reproduce:
- Given a file which fails with `ERROR_GENERIC`(sadly I cannot share my file).
- Call the `lz4.frame.decompress` over the file's bytes.
- The call will fail and the memory usage will increase, not decreasing afterwards.

After looking at the library's code, I wonder if the following line should include a call to `LZ4F_resetDecompressionContext`.
https://github.com/python-lz4/python-lz4/blob/54932796c78406ba24c8cfd0119ce56772fa0fb8/lz4libs/lz4frame.c#L1669

**Comment by jonathanunderwood** [comment]
Interesting. That line is in the [upstream lz4](https://github.com/lz4/lz4) code. Might be worth seeing if you can reproduce it with the lz4 tools supplied there.

**Comment by jmg-duarte** [comment]
I'm not sure how to replicate the leak using their tools, could you elaborate?

I did however tried to `unlz4` the file and got the same error but with error code 66 which led me to https://github.com/lz4/lz4/issues/560, any clue?

Below you can find a small demo of the issue, notice how the Python process memory consumption just gets higher and higher.

https://user-images.githubusercontent.com/15343819/161932311-49469a75-691b-40ed-b97d-789a932daa7f.mov

**Comment by jmg-duarte** [comment]
https://github.com/lz4/lz4/issues/560 is unrelated as discussed in https://github.com/lz4/lz4/issues/1074

**Comment by jmg-duarte** [comment]
Ok, I have updates.

I've written two scripts to test and theoretically, they should be equivalent (if not, let me know), however, one of them triggers the leak the other doesn't.
The following script does not leak memory.
```python
import lz4.frame
import os
import psutil

def mem_rss():
    print(psutil.Process(os.getpid()).memory_info().rss / 1024 ** 2)

print("initial")
mem_rss()

def read_lz4():
    try:
        with lz4.frame.open("error_generic.tar.lz4") as f:
            data = f.read()
    except Exception as e:
        print(e)

for i in range(10):
    read_lz4()
    print(i)
    mem_rss()
```
This one does:
```python
import lz4.frame
from lz4.frame import decompress
import os
import psutil

def mem_rss():
    print(psutil.Process(os.getpid()).memory_info().rss / 1024 ** 2)

print("initial")
mem_rss()

def read_lz4():
    with open("error_generic.tar.lz4", "rb") as f:
        data = f.read()
    try:
        decompressed_data = decompress(data)
    except Exception as e:
        print(e)

for i in range(10):
    read_lz4()
    print(i)
    mem_rss()
```

I've ran both scripts using `memray`, you can find the results bellow:

No leak:
![image](https://user-images.githubusercontent.com/15343819/176029279-a9a71fd0-9a73-45d1-95e9-2a15631c79bf.png)

Leak:
![image](https://user-images.githubusercontent.com/15343819/176029355-215310f9-6773-485d-953b-c81beb57688e.png)

Sadly, I still can't provide the file which triggers this.

**Comment by jonathanunderwood** [comment]
Fixed in 4.0.2.

**Comment by pablogsal** [comment]
👋 Hi @jmg-duarte,

I am one of the authors of `memray`. We are collecting success stories [here](https://github.com/bloomberg/memray/discussions/226). If you have a minute, do you mind leaving a short message on how `memray` helped with this issue? Knowing how we managed to help will help us track trends and target areas for improvement, prioritize new features and development and identify potential bugs or areas of confusion.

Thanks a lot for your consideration and for helping us improving the profiler :)
